### PR TITLE
9318 vol_volsize_to_reservation does not account for raidz skip blocks

### PR DESCRIPTION
**Pull request by mgerdts**
Reviewed by: Richard Elling <Richard.Elling@RichardElling.com>
Reviewed by: Sanjay Nadkarni <sanjay.nadkarni@nexenta.com>

### CODE REVIEW

File: usr/src/lib/libzfs/common/libzfs_dataset.c [hunk 1]
```diff
@@ -5111,12 +5113,177 @@ zfs_get_holds(zfs_handle_t *zhp, nvlist_t **nvl)
...
+ * to the amount of space required to store a 128k (SPA_OLD_MAXBLOCKSIZE) block.
+ * This is referred to as "raidz deflation."
+ *
+ * No matter how many blocks are allocated for a newly written uncompressed 128k
+ * block, referenced will increase by 128k.  However, the pool's allocated and
+ * free properties will be adjusted by the amount space allocated for data,
+ * parity, and perhaps skip blocks.  vdev_raidz_asize() is used for calculating
+ * the total amount of space required for a single block.  If another block size
+ * is written, the change to the referenced property may be different than the
+ * amount of space actually alloctaed.
```
Comment (ahrens) [inline usr/src/lib/libzfs/common/libzfs_dataset.c]: Great comment!  I think this one paragraph could be rewritten to emphasize the points that are salient here.  Here's a start:

> The "referenced" property of RAIDZ vdevs is scaled such that a 128KB block will "reference" 128KB, even though it allocates more than that, to store the parity information (and perhaps skip sectors).  This concept of the "referenced" (and other DMU space accounting) being lower than the allocated space by a constant factor is called "raidz deflation."
>
> As mentioned above, the constant factor for raidz deflation assumes a 128KB block size.  However, zvols typically have a much smaller block size (default 8KB).  These smaller blocks may require proportionally much more parity information (and perhaps skip sectors). In this case, the change to the "referenced" property may be much more than the logical block size.

File: usr/src/lib/libzfs/common/libzfs_dataset.c [hunk 2]
```diff
@@ -5111,12 +5113,177 @@ zfs_get_holds(zfs_handle_t *zhp, nvlist_t **nvl)
...
+ * Compression may lead to a variety of block sizes being written for the same
+ * volume or file.  There is no clear way to reserve just the amount of space
+ * that will be required, so the worst case (no compression) is assumed.
+ * Note that metadata blocks will typically be compressed, so the reservation
+ * size returned by zvol_volsize_to_reservation() will generally be slightly
+ * larger than the maximum that the volume can reference.
+ */
+
+/*
+ * Derived from function of same name in uts/common/fs/zfs/vdev_raidz.c
```
Comment (ahrens) [inline usr/src/lib/libzfs/common/libzfs_dataset.c]: how about adding: `Returns the amount of space (in bytes) that will be allocated for the specified block size. 
 Note that the "referenced" space accounted will be less than this, but not necessarily equal to "blksize", due to RAIDZ deflation.`

File: usr/src/lib/libzfs/common/libzfs_dataset.c (hunk 1 above)
Comment (mgerdts) [reply usr/src/lib/libzfs/common/libzfs_dataset.c]: Thanks, that looks great.

**Comment by mgerdts** [comment]
This has been integrated.

https://github.com/illumos/illumos-gate/commit/b73ccab03ec36581b1ae5945ef1fee1d06c79ccf

**Pull request closed without merging**